pkg/ingester/ingester.go

```diff
@@ -3137,10 +3142,40 @@ func (i *Ingester) Push(ctx context.Context, req *mimirpb.WriteRequest) (*mimirp
 		mimirpb.ReuseSlice(req.Timeseries)
 	})
 	err := i.PushWithCleanup(ctx, pushReq)
-	if err != nil {
-		return nil, err
+	if err == nil {
+		return &mimirpb.WriteResponse{}, nil
+	}
+	handledErr := handlePushError(err)
+	return nil, handledErr
+}
+
+func handlePushError(err error) error {
+	switch {
+	case errors.As(err, &unavailableError{}):
+		return newErrorWithStatus(
+			err,
+			codes.Unavailable,
+		)
+	case errors.As(err, &instanceLimitReachedError{}):
+		return newErrorWithStatus(
+			util_log.DoNotLogError{Err: err},
+			codes.Unavailable,
+		)
+	case errors.As(err, &tsdbUnavailableError{}):
+		return newErrorWithHTTPStatus(
+			err,
+			http.StatusServiceUnavailable,
+		)
+	case errors.As(err, &validationError{}) ||
+		errors.As(err, &perUserLimitReachedError{}) ||
+		errors.As(err, &perMetricLimitReachedError{}):
+		return newErrorWithHTTPStatus(
+			err,
+			http.StatusBadRequest,
+		)
```

Comment:
These are the most common ones, I'd put them as first branch to avoid having to check the rest (which are quite uncommon), and I'd add a method to them `interface { userError() }` and check that interface once here.

Reply:
I have made them implement the following interface:
```
type BadDataErr interface {
  badData()
}
```

Reply:
(Note that the interface doesn't need to be exported)